lib/src/piece/chain.dart

```diff
@@ -5,43 +5,205 @@ import '../back_end/code_writer.dart';
 import '../constants.dart';
 import 'piece.dart';
 
-// TODO(tall): This will probably become more elaborate when full method chains
-// with interesting argument lists are supported. Right now, it's just the
-// basics needed for instance creation expressions which may have method-like
-// `.` in them.
-
 /// A dotted series of property access or method calls, like:
 ///
 ///     target.getter.method().another.method();
 ///
-/// This piece handles splitting before the `.`.
+/// This piece handles splitting before the `.` and controlling which argument
+/// lists in the method calls are allowed to contain newlines.
+///
+/// Chains can split in four ways:
+///
+/// [State.unsplit] The entire chain on one line:
+///
+///     target.getter.method().another.method();
+///
+/// [_blockFormatTrailingCall] Don't split before any `.`. Split the last (or
+/// next-to-last if there is a hanging unsplittable call at the end) method
+/// call in the chain like a block while leaving other calls unsplit, as in:
+///
+///     target.property.first(1).block(
+///       argument,
+///       argument,
+///     );
+///
+/// [_splitAfterProperties] Split the call chain at each method call, but leave
+/// the leading properties on the same line as the target. We allow leading
+/// properties to remain unsplit while splitting the rest of the chain since
+/// property accesses often feel "closer" to the target then the methods called
+/// on it, as in:
+///
+///     motorcycle.wheels.front
+///         .rotate();
+///
+/// [State.split] Split before every `.` and indent the chain, like:
+///
+///     target
+///         .getter
+///         .method(
+///           argument,
+///           argument,
+///         ).another.method(
```

Comment:
Comment says before every `.`, but there is no split before the second `.method`.

Reply:
Oops, typo. :) Fixed.